server/src/main/java/org/elasticsearch/action/bulk/IncrementalBulkService.java

```diff
@@ -0,0 +1,194 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class IncrementalBulkService {
+
+    private final Client client;
+
+    public IncrementalBulkService(Client client) {
+        this.client = client;
+    }
+
+    public Handler newBulkRequest() {
+        return newBulkRequest(null, null, null);
+    }
+
+    public Handler newBulkRequest(@Nullable String waitForActiveShards, @Nullable TimeValue timeout, @Nullable String refresh) {
+        return new Handler(client, waitForActiveShards, timeout, refresh);
+    }
+
+    public static class Handler {
+
+        private final Client client;
+        private final ActiveShardCount waitForActiveShards;
+        private final TimeValue timeout;
+        private final String refresh;
+
+        private final ArrayList<Releasable> releasables = new ArrayList<>(4);
+        private final ArrayList<BulkResponse> responses = new ArrayList<>(2);
+        private boolean globalFailure = false;
+        private boolean incrementalRequestSubmitted = false;
+        private Exception topLevelFailure = null;
+        private BulkRequest bulkRequest = null;
+
+        private Handler(Client client, @Nullable String waitForActiveShards, @Nullable TimeValue timeout, @Nullable String refresh) {
+            this.client = client;
+            this.waitForActiveShards = waitForActiveShards != null ? ActiveShardCount.parseString(waitForActiveShards) : null;
+            this.timeout = timeout;
+            this.refresh = refresh;
+            createNewBulkRequest(BulkRequest.IncrementalState.EMPTY);
+        }
+
+        public void addItems(List<DocWriteRequest<?>> items, Releasable releasable, Runnable nextItems) {
+            if (topLevelFailure != null) {
+                assert releasables.isEmpty();
+                assert bulkRequest == null;
+                if (globalFailure == false) {
+                    addItemLevelFailures(items);
+                }
+                Releasables.close(releasable);
+                nextItems.run();
+            } else {
+                internalAddItems(items, releasable);
+
+                if (shouldBackOff()) {
+                    client.bulk(bulkRequest, ActionListener.runAfter(new ActionListener<>() {
+
+                        private final boolean isFirstRequest = incrementalRequestSubmitted == false;
+
+                        @Override
+                        public void onResponse(BulkResponse bulkResponse) {
+                            responses.add(bulkResponse);
+                            releaseCurrentReferences();
+                            createNewBulkRequest(bulkResponse.getIncrementalState());
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            handleBulkFailure(isFirstRequest, e);
+                        }
+                    }, nextItems));
+                    incrementalRequestSubmitted = true;
```

Comment:
Should we set this before invoke `client.bulk` (need to move out the `isFirstRequest` calculation? While this might work in a netty eventloop with single thread per request I wonder if we want to rely on that? I.e., once we invoke `nextItems.run()`, I could see this method being invoked in another thread (perhaps even same, in case the `bulk` request fails) as an option (though perhaps not in netty, but that seems subtle to rely on).